src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs

```diff
@@ -53,16 +53,20 @@ public partial interface IChangeToken
         public string? Value { get { throw null; } }
         public System.ReadOnlyMemory<char> AsMemory() { throw null; }
         public System.ReadOnlySpan<char> AsSpan() { throw null; }
+        [System.Diagnostics.CodeAnalysis.MemberNotNull(nameof(Buffer))]
+        [System.Diagnostics.CodeAnalysis.MemberNotNull(nameof(Value))]
```

Comment:
I question how much value these `MemberNotNull` attributes provides on these APIs. When would a caller:

1. Call AsSpan/Substring/Subsegment on a StringSegment
2. And then afterwards try to get the `Value` or `Buffer` of the StringSegment they got a piece of above

I would assume the call to AsSpan/Substring/Subsegment would already be guarded by a `HasValue` check, because if it wasn't, AsSpan/Substring/Subsegment would throw when `HasValue` is false.

This doesn't seem like a typical code flow. Do we have evidence in ASP.NET that this is actually beneficial? If so, can you provide a link to the real-world code?

Reply:
I just thought giving complier more information about the code is a good thing. But you are right, I don't think there would be many (if any) places where those attributes would save someone a redundant null check. 

Do you want them removed?

Reply:
> Do we have evidence in ASP.NET that this is actually beneficial?

I got nothing but it's possible our code uses `StringSegment.IsNullOrEmpty(...)` and `Length` or `HasValue` checks more than is common.

Reply:
If we could annotate the static `IsNullOrEmpty`, that would be great. But my understanding is that it isn't possible to annotate today.

> Do you want them removed?

For now, I'd say let's remove them. Adding them limits the kind of changes we could make in the future. For example, if someone called `AsSpan(0, 0)`, returning an empty span, even if the backing string is `null`, would be possible. With the proposed annotations, that wouldn't be possible since this is saying `Buffer` and `Value` are not `null`.

Reply:
> Adding them limits the kind of changes we could make in the future

Good point. Didn't think about that removing those attributes would be a breaking change. Removed them from the PR.

> But my understanding is that it isn't possible to annotate today.

Yeah, it doesn't seem to be. Currently attributes only can tell compiler about parameter, field or property not being null, but not method results (`StringSegment.ToArray` or implicit cast if `HasValue` is checked for example), nor members of parameter, field or property. 
Maybe it'll be possible some day. Is there an issue for that?

Reply:
> If we could annotate the static IsNullOrEmpty, that would be great. But my understanding is that it isn't possible to annotate today.

&hellip;

> Maybe it'll be possible some day. Is there an issue for that?

@jaredpar not sure who on your team owns the nullability attributes but could you point us in the right direction please❔ Or, does an issue already exist somewhere❔

The case here is a `struct` passed to a `static` method which `return`s `false` when a couple of nullable properties on the `struct` are non-`null`. It's both the intervening `struct` and the fact the method is `static` that make the existing attributes unhelpful.